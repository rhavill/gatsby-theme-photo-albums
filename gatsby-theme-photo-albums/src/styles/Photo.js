/** @jsx jsx */
import { jsx } from 'theme-ui'
import PropTypes from 'prop-types'

const Photo = ({children, height}) => 
  <div sx={{variant: 'photoPage', height}}>
    {children}
  </div>

Photo.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  height: PropTypes.number.isRequired,
}

export default Photo