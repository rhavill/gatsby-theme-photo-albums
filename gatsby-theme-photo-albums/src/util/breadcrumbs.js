import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs