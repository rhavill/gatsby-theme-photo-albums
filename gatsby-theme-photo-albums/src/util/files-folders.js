const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}