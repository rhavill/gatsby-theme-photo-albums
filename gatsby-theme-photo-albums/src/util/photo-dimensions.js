export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}