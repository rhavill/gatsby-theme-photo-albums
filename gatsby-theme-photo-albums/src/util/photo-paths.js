const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension, ensureLeadingAndTrailingSlash} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => { 
    let previousUrl = null
    let previousUrlWithPage = null
    return indexedForEach(
      (file, i) => {
        const pageNumber = Math.ceil((i + 1) / photosPerPage)
        photoPaths[file.url] = {
          url: getPhotoPathWithPage(pageNumber, file.url),
          previousUrl: previousUrlWithPage,
          nextUrl: null,
          parentUrl: appendPageNumberToUrl(pageNumber, file.parentUrl),
        }
        if (previousUrl) {
          photoPaths[previousUrl].nextUrl = photoPaths[file.url].url
        }
        previousUrl = file.url
        previousUrlWithPage = photoPaths[file.url].url
      }, 
      groupedFiles[relativeDirectory]
    )
  }, keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `page-${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

const appendPageNumberToUrl = (pageNumber, url) => {
  let path = url ? url : null
  if (path && pageNumber > 1) {
    path = `${ensureLeadingAndTrailingSlash(path)}page-${pageNumber}`
  }
  return path
}

module.exports = {
  getPhotoPathsWithPages, 
  getPhotoPathWithPage,
}