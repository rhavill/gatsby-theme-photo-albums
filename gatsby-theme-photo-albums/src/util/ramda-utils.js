const addIndex = require('ramda/src/addIndex')
const complement = require('ramda/src/complement')
const compose = require('ramda/src/compose')
const forEach = require('ramda/src/forEach')
const groupBy = require('ramda/src/groupBy')
const isNil = require('ramda/src/isNil')
const map = require('ramda/src/map')
const prop = require('ramda/src/prop')

/**
 * Accepts an array of objects and returns an array containing the values of a 
 * single property from each object of the input array.
 *
 * @func
 * @sig Array -> Array
 * @param {String} property The property to extract from each object of the 
 *    input array.
 * @param {Array} objects The input array containing objects.
 * @return {Array} A simple array containing values exracted from the `property` 
 *    values of the input array.
 * @example
 *
 *      const objects = [{name: 'Hilda', age: 33}, {name: 'José', age: 45}]
 *      objectArrayToPropArray('name', objects); //=> ['Hilda', 'José']
 */
const objectArrayToPropArray = (property, objects) => map(prop(property), objects)

const mapIndexed = addIndex(map)

const indexedForEach = addIndex(forEach)

const groupByProp = property => compose(groupBy, prop)(property)

const isNotNil = complement(isNil)

module.exports = {
  objectArrayToPropArray,
  mapIndexed,
  indexedForEach,
  groupByProp,
  isNotNil,
}
