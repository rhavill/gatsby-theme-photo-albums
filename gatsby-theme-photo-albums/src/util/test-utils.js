import mergeDeepRight from 'ramda/src/mergeDeepRight'
import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'emotion-theming'

import theme from '../theme'

const create = (element, options) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{element}</ThemeProvider>, options
  )
}

const customRenderer = mergeDeepRight(renderer, {create})

const Wrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}
Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options })

// re-export everything from React Testing Library
export * from '@testing-library/react'

// override react testing library render method and custom react-test-renderer r
export { customRender as render, customRenderer as renderer }