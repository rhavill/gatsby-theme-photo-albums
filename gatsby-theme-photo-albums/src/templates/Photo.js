import both from 'ramda/src/both'
import compose from 'ramda/src/compose'
import find from 'ramda/src/find'
import prop from 'ramda/src/prop'
import propEq from 'ramda/src/propEq'
import React, {useState} from 'react'
import PropTypes from 'prop-types'
import {graphql, navigate, Link} from 'gatsby'
import Img from 'gatsby-image'
import Layout from '../components/layout/Layout'
import PhotoStyles from '../styles/Photo'
import {isNotNil} from '../util/ramda-utils'
import {pathToFileTitle, removePathPrefix, removeFileExtension} from '../util/url-text'
import useKeyUp from '../hooks/use-key-up'
import useSwipeNavigate from '../hooks/use-swipe-navigate'
import { TransformWrapper, TransformComponent } from 'react-zoom-pan-pinch'
import useWindowDimensions from '../hooks/use-window-dimensions'
import photoDimensions from '../util/photo-dimensions'

const Photo = ({data, path, pageContext}) => {
  const pathPrefix = data.site.pathPrefix
  path = compose(
    removeFileExtension, removePathPrefix(pathPrefix), decodeURI
  )(path)
  const title = pathToFileTitle(path)
  const windowDimensions = useWindowDimensions()
  const dimensions = photoDimensions(windowDimensions, pageContext)
  const previousUrl = pageContext.file.previousUrl
  const nextUrl = pageContext.file.nextUrl
  const parentUrl = pageContext.file.parentUrl
  const keyUpActions = [
    {navigateTo: parentUrl, wasPressed: useKeyUp('ArrowUp')},
    {navigateTo: previousUrl, wasPressed: useKeyUp('ArrowLeft')},
    {navigateTo: nextUrl, wasPressed: useKeyUp('ArrowRight')},
  ]
  const keyboardAction = find(
    both(
      compose(isNotNil, prop('navigateTo')),
      propEq('wasPressed', true)
    ),
    keyUpActions
  )
  if (keyboardAction) {
    navigate(keyboardAction.navigateTo)
  }
  const [swipeDisabled, setSwipeDisabled] = useState(false)
  const panDisabled = !swipeDisabled
  const swipeNavigation = useSwipeNavigate(nextUrl, previousUrl, parentUrl)
  const swipeHandlers = swipeDisabled ? {} : swipeNavigation
  const handleZoomChange = zoom => {
    if (zoom.scale === 1) {
      setSwipeDisabled(false)
    }
    else {
      setSwipeDisabled(true)
    }
  }
  
  return (
    <Layout path={path}>
      <div data-testid={path} {...swipeHandlers}>
        <PhotoStyles height={windowDimensions.height}>
          {previousUrl ?
            <div className='photo-navigation previous'>
              <div><Link to={previousUrl}>◄</Link></div>
            </div> 
            : null}
          <TransformWrapper onZoomChange={handleZoomChange} pan={{disabled: panDisabled}} doubleClick={{mode: 'reset'}}>
            <TransformComponent>
              <Img fixed={data.photo.childImageSharp.fixed} alt={title} title={title}
                loading='eager' css={dimensions} draggable={false} />
            </TransformComponent>
          </TransformWrapper>
          {parentUrl ?
            <div className='photo-navigation parent'>
              <div><Link to={parentUrl}>⯅</Link></div>
            </div>
            : null}
          {nextUrl ?
            <div className='photo-navigation next'>
              <div><Link to={nextUrl}>►</Link></div>
            </div>
            : null}
        </PhotoStyles>
      </div>
    </Layout>
  )
}

export const query = graphql`
  query photoQuery($relativePath: String!, $width: Int!, $height: Int!) {
    site {
      pathPrefix
    }
    photo: file(relativePath: {eq: $relativePath}) {
      childImageSharp {
        fixed(width: $width, height: $height) {
          ...GatsbyImageSharpFixed
        }
      }
    }
  }
`

Photo.propTypes = {
  path: PropTypes.string.isRequired,
  data: PropTypes.shape({
    site: PropTypes.shape({
      pathPrefix: PropTypes.string.isRequired,
    }),
    photo: PropTypes.shape({
      childImageSharp: PropTypes.shape({
        fixed: PropTypes.shape({
          src: PropTypes.string.isRequired,
          srcSet: PropTypes.string.isRequired,
          width: PropTypes.number.isRequired,
          height: PropTypes.number.isRequired,
        }).isRequired
      })
    })
  }),
  pageContext: PropTypes.shape({
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    file: PropTypes.shape({
      previousUrl: PropTypes.string,
      nextUrl: PropTypes.string,
      parentUrl: PropTypes.string,
    }),
  }).isRequired
}

export default Photo