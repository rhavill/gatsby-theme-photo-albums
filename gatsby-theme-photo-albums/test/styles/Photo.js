import React from 'react'
import {cleanup, renderer} from '../../src/util/test-utils'
import Photo from '../../src/styles/Photo'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Photo styles', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Photo height={600}><div>hello</div></Photo>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
