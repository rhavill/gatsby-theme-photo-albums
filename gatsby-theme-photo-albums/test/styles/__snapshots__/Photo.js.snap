// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo styles renders correctly 1`] = `
<div
  className="css-1oyu1ai-Photo"
>
  <div>
    hello
  </div>
</div>
`;
